arch/riscv/core/cache.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright 2023 Andestech
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/cache.h>
+#include <zephyr/sys/barrier.h>
+
+static struct k_spinlock lock;
+
+#ifdef CONFIG_DCACHE
+
+extern size_t __soc_get_dcache_line_size(void);
+extern void __soc_dcache_enable(void);
+extern void __soc_dcache_disable(void);
+size_t __weak __soc_get_dcache_line_size(void) {}
+void __weak __soc_dcache_enable(void) {}
+void __weak __soc_dcache_disable(void) {}
+
+size_t arch_dcache_line_size_get(void)
+{
+	size_t dcache_line_size = 0;
+
+	if (__soc_get_dcache_line_size) {
+		dcache_line_size = __soc_get_dcache_line_size();
+	}
+
+	return dcache_line_size;
+}
+
+void arch_dcache_enable(void)
+{
+	if (__soc_dcache_enable) {
+		__soc_dcache_enable();
+	}
+}
+
+void arch_dcache_disable(void)
+{
+	if (__soc_dcache_disable) {
+		__soc_dcache_disable();
+	}
+}
+
+int arch_dcache_flush_all(void)
+{
+	return -ENOTSUP;
+}
+
+int arch_dcache_invd_all(void)
+{
+	return -ENOTSUP;
+}
+
+int arch_dcache_flush_and_invd_all(void)
+{
+	return -ENOTSUP;
+}
+
+#ifndef CONFIG_CBOM_DCACHE_SUPPORTED
+int arch_dcache_flush_range(void *start_addr, size_t size)
+{
+	return -ENOTSUP;
+}
+
+int arch_dcache_invd_range(void *start_addr, size_t size)
+{
+	return -ENOTSUP;
+}
+
+int arch_dcache_flush_and_invd_range(void *start_addr, size_t size)
+{
+	return -ENOTSUP;
+}
+
+#else
```

Comment:
as per my previous comment you are implementing cache operations when `!CONFIG_{I,D}CACHE` and this is wrong.

Reply:
Thanks for your replying.
Sorry for that I didn't realise your comment exactly.
Because the else is not for the !CONFIG_{I,D}CACHE, is for !CONFIG_CBOM_{I,D}CACHE_SUPPORTED.
Our means is that, also when {I,D}CACHE is enable but not support CBOM_{I,D}CACHE_SUPPORTED, the api will return NOTSUP

Reply:
Thank you for your information.
Ok, I got it, I will use the inline function.

About the __soc_* functions,  I got what you mean, and it is actually insightful.
But I'd also like to discuss with you that our consideration is that the CMO isn't supporting all cache functions.
So if we provide a hook function to the SoC layer, there is an opportunity to enhance the functionality of the arch cache API. 
For example, if without having a hook function for the SoC,  the cache line size can't dynamic retrieval in functions like cache flush, invalidation, and flush & invalidate.
The only option to get line size would be relying on the device tree(dts node info) or Kconfig, which means getting the value offline.
We're not sure if our consideration at this is makes sense or not.

Reply:
I think both my intended goals.
Because if we aim to dynamically obtain the cache line size during Cache range operations, it is indeed necessary for the SoC to provide a hook function in order to accomplish this.
And due to the not-supported of cache enable/disable and obtaining line size in CMO standard extensions, we think that others may encounter similar issues. Therefore, we propose using a hook function as a solution.